### PR TITLE
Sparse gradient update in the SGD optimizer

### PR DESCRIPTION
**Pull request by entelechie**
Support sparse gradients in SGD to improve embedding layers training. Related issue: [4365](https://github.com/fchollet/keras/issues/4365)

IMPACT:
- [TENSORFLOW, EMBEDDING] Handle sparse gradients from embedding layers. For a model with 409616384 parameters in its embedding layer, the times per batch decreases from 86ms to 33ms for SGD without momentum. The times per batch decreases from 103ms to 66ms for SGD with momentum. I was expecting the momentum will cripple sparse updates. But it looks like even sparse updating on the moment variable could help.
- [TENSORFLOW] Replace several intermediate tensors with in-place updates.
- [GENRALLY] Avoid a NOOP with the zero moment variable when momentum==0. The moment variable is big and expensive to evaluate. Vanilla SGD might run faster.

I didn't copy Tensorflow's apply_sparse() and apply_dense() interface because IMO it could be painful to keep two implementations consistency in the long term.

### CODE REVIEW

**Comment by zippeurfou** [comment]
@entelechie What happened with this PR? 
It seems that https://github.com/keras-team/keras/issues/4365 still exist no?
